python/cog/types.py

```diff
@@ -131,17 +155,44 @@ def unlink(self, missing_ok: bool = False) -> None:
                 if not missing_ok:
                     raise
 
+
+class DataURLTempFilePath(pathlib.PosixPath):
+    def __init__(self, url: str) -> None:
+        resp = urllib.request.urlopen(url)  # noqa: S310
+        self.source = get_filename_from_urlopen(resp)
+        dest = tempfile.NamedTemporaryFile(suffix=self.source, delete=False)
+        shutil.copyfileobj(resp, dest)
+        self._path = pathlib.Path(dest.name)
+
+    def convert(self) -> pathlib.Path:
+        return self._path
+
     def __str__(self) -> str:
         # FastAPI's jsonable_encoder will encode subclasses of pathlib.Path by
         # calling str() on them
         return self.source
 
+    def unlink(self, missing_ok: bool = False) -> None:
+        if self._path:
+            # TODO: use unlink(missing_ok=...) when we drop Python 3.7 support.
+            try:
+                self._path.unlink()
+            except FileNotFoundError:
+                if not missing_ok:
+                    raise
+
+
+# we would prefer URLFile to stay lazy
+# except... that doesn't really work with httpx?
+
 
 class URLFile(io.IOBase):
     """
     URLFile is a proxy object for a :class:`urllib3.response.HTTPResponse`
     object that is created lazily. It's a file-like object constructed from a
     URL that can survive pickling/unpickling.
+
+    This is the only place Cog uses requests
```

Comment:
What would it take to get rid of requests outright?

Would something like this work?

```python
@property
def __wrapped__(self) -> Any:
    try:
        return object.__getattribute__(self, "__target__")
    except AttributeError:
        url = object.__getattribute__(self, "__url__")
        with httpx.stream("GET", url) as resp:
            resp.raise_for_status()
            resp.raw.decode_content = True
            object.__setattr__(self, "__target__", resp.raw)
            return resp.raw
```

Reply:
I think that would work, I just didn't bother yet because nobody anywhere uses File and this solution is still a little unsatisfying (it would block the event loop). I also suspect requests is always installed because pip depends on it.

Reply:
> requests is always installed because pip depends on it 

🤨 

Maybe that's no longer the case? According to [this](https://libraries.io/pypi/pip), pip has no external dependencies.

Not urgent or blocking, but it'd be nice to make a clean break.

Reply:
oh yeah they vendored it in a while ago, I forgot. and I'm mistaken, python:3.11-slim doesn't have requests by default. 

I'd still lean towards leaving it obviously broken rather then seeming okay but actually blocking the event loop. good ways to do this include subprocesses with a pipe, some kind of thread nonsense or a hypothetical pget-py binding